File: src/main/java/artisra/leetcode75/arraystring/GreatestCommonDivisorOfStrings.java
package artisra.leetcode75.arraystring;

public class GreatestCommonDivisorOfStrings {
    public String gcdOfStrings(String str1, String str2) {
        if (!(str1 + str2).equals(str2 + str1)) return "";
        int strGcdLength = gcd(str1.length(), str2.length());
        return str1.substring(0, strGcdLength);
    }

    private int gcd(int a, int b) {
        while (b != 0) {
            int remainder = a % b;
            a = b;
            b = remainder;
        }
        return a;
    }
}

File: src/main/kotlin/leetcode75/arraystring/GreatestCommonDivisorOfStrings.kt
package dev.artisra.leetcode75.arraystring

class GreatestCommonDivisorOfStrings {

    fun gcdOfStrings(str1: String, str2: String): String {
        if (str1 + str2 != str2 + str1) return ""
        val length = gcd(str1.length, str2.length)
        return str1.substring(0 until length)
    }

    private fun gcd(a: Int, b: Int): Int {
        var x = a
        var y = b
        while (y != 0) x = y.also { y = x % y }
        return x
    }
}